x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportGetShutdownStatusAction.java

```diff
@@ -195,7 +198,47 @@ static ShutdownShardMigrationStatus shardMigrationStatus(
             return new ShutdownShardMigrationStatus(SingleNodeShutdownMetadata.Status.COMPLETE, 0);
         }
 
-        // First, check if there are any shards currently on this node, and if there are any relocating shards
+        final RoutingAllocation allocation = new RoutingAllocation(
+            allocationDeciders,
+            currentState,
+            clusterInfoService.getClusterInfo(),
+            snapshotsInfoService.snapshotShardSizes(),
+            System.nanoTime()
+        );
+        allocation.setDebugMode(RoutingAllocation.DebugMode.EXCLUDE_YES_DECISIONS);
+
+        // We also need the set of node IDs which are currently shutting down.
+        Set<String> shuttingDownNodes = currentState.metadata().nodeShutdowns().keySet();
+
+        // Check if we have any unassigned primary shards that have this nodeId as their lastAllocatedNodeId
+        var unassignedShards = Stream.iterate(
+            currentState.getRoutingNodes().unassigned().iterator(),
+            Iterator::hasNext,
+            UnaryOperator.identity()
+        ).map(Iterator::next).filter(s -> Objects.equals(s.unassignedInfo().getLastAllocatedNodeId(), nodeId)).toList();
+
+        if (unassignedShards.isEmpty() == false) {
+            for (var shardRouting : unassignedShards) {
+                if (shardRouting.primary() || hasShardCopyOnAnotherNode(currentState, shardRouting, shuttingDownNodes) == false) {
+                    ShardAllocationDecision decision = allocationService.explainShardAllocation(shardRouting, allocation);
+
+                    return new ShutdownShardMigrationStatus(
+                        SingleNodeShutdownMetadata.Status.STALLED,
+                        unassignedShards.size(),
```

Comment:
I think this number is misleading. It shows all of the unassigned shards but not all of them might stall the migration.
I believe it is possible to move `shardRouting.primary() || hasShardCopyOnAnotherNode(currentState, shardRouting, shuttingDownNodes) == false` check to the stream above. This way we would only have a list of shards that are stalling the migration.

Reply:
Good idea, I'll add another filter and simplify that check.